### PR TITLE
fix(moreh): prevent nan in norm_backward for zero-norm slices (#55585)

### LINKED ISSUE #55585: [Bug] moreh_norm_backward returns NaN for all-zero reduced slices, where the gradient is finite
## Summary

`ttnn.operations.moreh.norm_backward` computes the gradient of `norm(x, p, dim)` as

```
dx = sign(x) * |x|^(p-1) * y * dy * (1 / y^p)        where y = norm(x, p, dim)
```

with no guard for `y == 0`. When a reduced slice is entirely zero its norm is
zero, `1 / y^p` is non-finite, and it multiplies a zero factor, so every element
of that slice comes back non-finite. PyTorch returns a finite `0`.

```
input:    [[0, 0, 0],
           [1, 2, 2]]
p:        2.0
dim:      1
grad:     [1, 1]

norm:     [0, 3]
expected: [[0,      0,      0     ],   <- PyTorch
           [0.3333, 0.6667, 0.6667]]
current:  [[non-finite, non-finite, non-finite],
           [0.3333,     0.6667,     0.6667    ]]
```

An all-zero reduced slice is not exotic: zero-initialised embedding rows, padded
or fully-masked sequence positions, and dead channels all produce one. A single
non-finite gradient propagates through the optimiser and destroys the whole
parameter tensor on the next step.

The same missing guard also affects `p < 1` with an exactly-zero *element*, even
when the slice's norm is non-zero: `|0|^(p-1)` is non-finite there, and PyTorch
masks that case too (see below).

## Root cause

`ttnn/cpp/ttnn/operations/moreh/moreh_norm_backward/device/kernels/moreh_norm_backward_kernel.cpp`

```
L50   sign  = sign_tile(x)
L53   xpow  = |x|^(p-1)          power_tile_with_abs_x_to_cb
L63   tmp4  = xpow * y
L90   tmp5  = tmp4 * dy
L118  recip = 1 / y^p            power_and_recip_tile_to_cb
L128  tmp4  = tmp5 * recip
L160  dx    = sign * tmp4        mul_tiles_to_cb
```

`power_and_recip_tile_to_cb`
(`ttnn/cpp/ttnn/kernel/compute/moreh_common.hpp`) ends in a `recip_tile` on
`y^p`. For `y == 0` that reciprocal is non-finite, and the surrounding factors
(`xpow`, `y`, and `sign` — all zero when the slice is all zeros) are exactly the
zeros that turn it into a NaN rather than cancelling it.

There is no `y == 0` guard anywhere on the path:

- the compute kernel above,
- `.../device/moreh_norm_backward_program_factory.cpp`,
- `.../device/moreh_norm_backward_device_operation.cpp`,
- `.../moreh_norm_backward.cpp` (host entry, forwards straight to `ttnn::prim`).

PyTorch derives the same formula but zeroes the singular cases explicitly.
`torch/csrc/autograd/FunctionsManual.cpp`, `norm_backward` (quoting v2.11.0,
the version pinned in `tt_metal/python_env/requirements-dev.txt`):

```cpp
} else if (p == 2.0) {
    return grad * (self / norm).masked_fill_(norm == 0, 0);
} else if (p < 1.0) {
    self_scaled = self.sgn() * self.abs().pow_(p - 1).masked_fill_(self == 0, 0);
    return self_scaled * grad * norm.pow(1 - p);
} else if (p < 2.0) {
    self_scaled = self.sgn() * self.abs().pow_(p - 1);
    scale_v = grad / norm.pow(p - 1);
    scale_v.masked_fill_(norm == 0, 0);
    return self_scaled * scale_v;
} else {
    self_scaled = self * self.abs().pow_(p - 2);
    scale_v = grad / norm.pow(p - 1);
    scale_v.masked_fill_(norm == 0, 0);
    ...
```

Every branch carries a `masked_fill_`. That is the piece tt-metal inlined away.

Unlike a reduction that loses information (the `prod_bw` zero case, for
instance), nothing here is unrecoverable — the correct answer is simply `0`, and
the formula fails only because `0 * inf` is NaN. So the fix is cheap. Two
caveats on how to apply it:

- Clamping `y` to some `eps` before the reciprocal is enough for `p >= 1`, since
  the `|x|^(p-1)` factor is itself `0` there and kills the large reciprocal.
- It is *not* enough for `p < 1`, where `|0|^(p-1)` is itself non-finite and no
  clamp on `y` can rescue it. That branch needs the element-wise
  `where(x == 0, 0, ...)` that PyTorch applies.

An explicit `where(norm == 0, 0, dx)` covers both and mirrors PyTorch directly,
which also keeps the intent legible: `0` here is the sub-gradient chosen by
convention at a point where the norm is not differentiable, not a limit of the
formula.

For contrast, `moreh_clip_grad_norm` in this same family *does* carry the
corresponding guard (`moreh_clip_grad_norm.cpp:126-129`, `max_norm / (total_norm
+ 1e-6)`), which suggests the omission in `norm_backward` is an oversight rather
than a deliberate contract.

## Reproducer

I do not have a card. This is a source-level analysis: the script below replays
the kernel's arithmetic on the host in fp32 and compares it against PyTorch
autograd. It does not exercise the device.

```
pip install "torch==2.11.0"
python repro_moreh_norm_backward.py
```

```python
import numpy as np, torch
f32 = np.float32

def ttnn_moreh_norm_backward(x, y, dy, p):
    """Replay of moreh_norm_backward_kernel.cpp in fp32.
    x: input (..., N)   y: forward norm, keepdim (..., 1)   dy: grad (..., 1)"""
    x, y, dy = x.astype(f32), y.astype(f32), dy.astype(f32)
    with np.errstate(divide="ignore", invalid="ignore", over="ignore"):
        sign = np.sign(x).astype(f32)
        xpow = np.power(np.abs(x), f32(p - 1.0)).astype(f32)   # |x|^(p-1)
        tmp4 = (xpow * y).astype(f32)
        tmp5 = (tmp4 * dy).astype(f32)
        recip = (f32(1.0) / np.power(y, f32(p))).astype(f32)   # 1 / y^p
        return (sign * (tmp5 * recip)).astype(f32)

x  = np.array([[0., 0., 0.], [1., 2., 2.], [0., 3., 4.]], dtype=f32)
for p in (0.5, 1.0, 2.0, 2.5, 3.0):
    y  = np.linalg.norm(x.astype(np.float64), ord=p, axis=1, keepdims=True).astype(f32)
    dy = np.ones_like(y)
    tt = ttnn_moreh_norm_backward(x, y, dy, p)

    t = torch.tensor(x, requires_grad=True)
    torch.linalg.vector_norm(t, ord=p, dim=1, keepdim=True).backward(torch.tensor(dy))
    ref = t.grad.numpy()

    print(f"p={p:<5} ttnn non-finite: {int((~np.isfinite(tt)).sum())}  "
          f"torch non-finite: {int((~np.isfinite(ref)).sum())}")
```

Output (torch 2.11.0+cpu):

```
p=0.5   ttnn non-finite: 4  torch non-finite: 0
p=1.0   ttnn non-finite: 3  torch non-finite: 0
p=2.0   ttnn non-finite: 3  torch non-finite: 0
p=2.5   ttnn non-finite: 3  torch non-finite: 0
p=3.0   ttnn non-finite: 3  torch non-finite: 0
```

The three are the all-zero row. The fourth at `p = 0.5` is the exact zero in row
`[0, 3, 4]`, which is the `masked_fill_(self == 0, 0)` branch PyTorch takes for
`p < 1`.

Per-element table for `p = 2.0`:

```
input row    norm    ttnn dx                    torch dx
[0, 0, 0]    0       [nan, nan, nan]            [0, 0, 0]          <- diverges
[1, 2, 2]    3       [0.3333, 0.6667, 0.6667]   [same]             ok
[0, 3, 4]    5       [0, 0.6, 0.8]              [same]             ok
```

Two things I could **not** verify without hardware, stated as open questions
rather than claims:

1. Whether the SFPU yields NaN or ±inf here. The host replay gives NaN. Either
   way the result is non-finite where PyTorch is finite, since nothing on the
   path can produce `0`; but I cannot confirm which.
2. `power_tile_with_abs_x_to_cb` evaluates `exp(log|x| * decimal)` unconditionally,
   and `decimal` is `0` for integer `p`. If the SFPU's `log(0)` returns `-inf`,
   then `-inf * 0` is NaN and `|x|^(p-1)` would be NaN at *every* exactly-zero
   element, not only in all-zero slices — widening the blast radius considerably.
   If `log_tile` instead saturates to a finite negative value, this second effect
   does not occur and only the all-zero-slice case above applies. I have no way
   to tell which from the source alone.

## Additional observation: `p = ±inf` on the backward path

Not the subject of this report, but it surfaced while reading the same code and
seems worth recording here rather than in a separate issue.

`moreh_norm_backward_program_factory.cpp:37-45` decomposes `p` with

```cpp
std::tuple<uint32_t, float, bool> get_floored_p_and_decimal_and_p_is_negative(float p) {
    auto floored_p = std::floor(p);
    auto decimal = p - floored_p;
    ...
    return std::make_tuple(static_cast<uint32_t>(floored_p), decimal, p_is_negative);
}
```

For `p = ±inf` this evaluates `static_cast<uint32_t>(std::floor(inf))`, which is
undefined behaviour for an out-of-range floating-to-integer conversion, and
`decimal = inf - inf = NaN`. There is also no `isinf(p)` branch in the compute
kernel, whereas PyTorch's `norm_backward` has a dedicated one (the `argmax` mask
over `self.abs() == norm`).

The forward test parametrises `p` over `[2.0, 2.5, -2.5, 0.0, inf, -inf]`, but
the backward test
(`test_moreh_norm.py`, `test_moreh_norm_backward`) only covers
`[2.0, 2.5, -2.5]`, so neither `±inf` nor `0.0` is exercised on this path. For
reference, torch 2.11.0 returns finite gradients for `p = ±inf` and no gradient
at all for `p = 0` (`norm_backward` returns `{}`). I have not tried to
characterise what the current kernel does for these values — I only want to flag
that they appear to be unhandled and untested.

## Expected behavior

Match `norm_backward`: where the norm of a slice is zero, the gradient for every
element of that slice is `0`.

- **no zeros in the slice** — current behaviour is already correct.
- **some zeros, norm > 0** — the host replay comes out correct for `p >= 1`
  (`sign(0) = 0` absorbs the result), though see open question 2 above: on device
  this depends on what `log_tile(0)` returns. For `p < 1`, elements that are
  exactly zero should be forced to `0` (PyTorch's
  `.masked_fill_(self == 0, 0)`), not `|0|^(p-1)`.
- **whole slice zero, norm == 0** — every element of the slice should be `0`.
  Currently non-finite.

Concretely, the kernel needs the equivalent of `where(y == 0, 0, dx)` applied
after the reciprocal-multiply, broadcast over the reduced dimension, plus a
`where(x == 0, 0, xpow)` on the `p < 1` path.

## Suggested tests

`tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_norm.py` builds its
inputs with `make_torch_tensors`:

```python
torch_input = torch.empty(input_shape, dtype=dtype).uniform_(-1, 1).requires_grad_()
```

A continuous uniform draw makes an all-zero reduced slice a probability-zero
event, so `norm == 0` is never exercised — which is why this has stayed hidden.
Worth adding to `run_moreh_norm_backward`:

1. An input with at least one all-zero slice along the reduced dim, asserting the
   gradient is finite and exactly zero there, for `p` in
   `{0.5, 1.0, 2.0, 2.5, 3.0, inf}`.
2. An all-zero input tensor (every slice degenerate).
3. For `p < 1`, a slice mixing exact zeros with non-zeros, asserting the zero
   positions get `0` rather than a non-finite value.
4. A `torch.isfinite(...).all()` assertion in `run_moreh_norm_backward`
   generally — `comp_allclose` currently compares values but the existing
   generator never produces the degenerate inputs that would trip it.

## One question before I go further

I would like to pick this up and submit a fix, and I understand I need to be
assigned first.

I noticed #46147 proposes documenting moreh ops as deprecated. Before I invest
in a fix, could you confirm whether `moreh_norm_backward` is still in scope for
the bounty program?

Happy to write the patch and the tests above as soon as I hear back.

### PR DESCRIPTION
**Pull request by mohmedmm**
fixes #55585

Replaces the final multiplication in moreh_norm_backward_kernel with mul_tiles_and_mask_tile_to_cb keyed on dfb_sign_obj. This eliminates non-finite gradients at x == 0 and for reduced zero-norm slices without relying on epsilon headroom. Extends unit tests to cover p in [0.5, 5.0] across all-zero and mixed-zero slices against PyTorch autograd.

### CODE REVIEW

**Comment by chonglokfong-ops** [comment]
Two cases from #55585 that I don't think this covers yet. I don't have a card,
so everything below is source-level analysis plus a host-side fp32/bf16 replay of
the kernel arithmetic — happy to be corrected if the device behaves differently.

### 1. The epsilon's headroom runs out around `p ≈ 3.2`, and the test stops at `p = 3.0`

`power_and_recip_tile_to_cb()`
(`ttnn/cpp/ttnn/kernel/compute/moreh_common.hpp:1077-1168`) evaluates `y^p` in
full — `power_iterative_tile()` on the integer part, times `exp(log(y) * decimal)`
— and only then calls `recip_tile()`. So `1 / y^p` has to be representable as a
value, it is never formed as a fused division.

With `y = 0 + 1e-12`:

| `p` | `y^p` (fp32) | `1 / y^p` (fp32) | covered by the new test? |
|---|---|---|---|
| 2.0 | 1.00e-24 | 1.00e+24 | yes |
| 2.5 | 1.00e-30 | 1.00e+30 | yes |
| 3.0 | 1.00e-36 | 1.00e+36 | yes |
| 3.2 | 3.98e-39 | 2.51e+38 | no |
| 3.5 | 1.00e-42 | overflows | no |
| 4.0 | 0 (below fp32's smallest subnormal; in bf16 it underflows sooner) | — | no |

fp32 and bf16 share an 8-bit exponent, so both cap at `max ≈ 3.4e38` and the
cutoff is the same for either intermediate format: `1 / y^p` stops being
representable once `y^p < 1 / max ≈ 2.9e-39`, i.e. once `p > 3.21`.

The added test parametrises `p ∈ [1.0, 2.0, 2.5, 3.0]`, which validates up to
3.0 and never crosses that boundary. `p` is not constrained anywhere on this path
— `MorehNormBackwardOperation::validate_inputs`
(`device/moreh_norm_backward_device_operation.cpp:14-20`) only calls
`check_tensor` on the four tensors, and the nanobind arg is a plain float — so
`p = 4` is reachable from Python.

I can't tell from the source what `recip_tile()` does on a denormal or zero
input, so I'm not claiming a specific output value past `p > 3.21`; the point is
that the working range is set by the magnitude of the epsilon rather than by
anything intrinsic to the op, and the tests stop just inside it.

### 2. `p < 1` with a zero element in a slice whose norm is non-zero

This one is independent of the norm, so adding an epsilon to `output` doesn't
reach it.

For `p < 1` the exponent `p - 1` is negative, and
`power_tile_with_abs_x_to_cb()` (`moreh_common.hpp:979-1076`) evaluates
`|x|^(p-1)` as `recip(|x|^floor)` times `exp(log|x| * decimal)`. At `x == 0` that
is `recip(0)` multiplied by `exp(-inf * decimal)`. The true limit of `|x|^(p-1)`
as `x → 0` is `+inf`, so there is no value the SFPU can return here that gives
the right answer — the case needs a guard, not a better reciprocal.

Concrete example, `p = 0.5`, one row, replayed in fp32 on the host:

```
input           [0.0, 3.0, 4.0]     p = 0.5
norm            13.928204           <- non-zero, so the epsilon never applies

     x      |x|^(p-1)          y^p        1/y^p           dx          torch 2.11
   0.0            nan       3.7321       0.2679          nan            0.000000
   3.0         0.5774       3.7321       0.2679       2.1547            2.154701
   4.0         0.5000       3.7321       0.2679       1.8660            1.866025
```

torch 2.11.0 handles this in a separate branch — the `p < 1.0` branch of
`norm_backward` in `torch/csrc/autograd/FunctionsManual.cpp`:

```cpp
} else if (p < 1.0) {
    self_scaled =
        self.sgn() * self.abs().pow_(p - 1).masked_fill_(self == 0, 0);
    return self_scaled * grad * norm.pow(1 - p);
}
```

Two things there: the guard is on `self == 0`, not on `norm == 0`; and it
multiplies by `norm.pow(1 - p)` rather than dividing, so this branch never forms
`1 / y^p` at all. The new test's lowest `p` is 1.0, so `p < 1` isn't exercised.

### Suggestion

Rather than an epsilon, it might be worth mirroring torch's branch structure,
since the per-`p` branch dispatch in `norm_backward` (torch 2.11.0) uses a
different guard per range of `p`:

- `p == 0` → no gradient
- `p == 1` → `sgn(self) * grad`, no division at all
- `p == 2`, `1 < p < 2`, `p >= 2` → divide, then `masked_fill_(norm == 0, 0)`
- `p < 1` → `masked_fill_(self == 0, 0)` on the element term, multiply by `norm^(1-p)`
- `isinf(p)` → argmax mask over `self.abs() == norm`

The compute kernel already has both the `x` and `y` tiles in scope and already
handles the broadcast cases, so an explicit `where(y == 0, 0, dx)` at the end —
plus `where(x == 0, 0, dx)` when `p < 1` — would give exact zeros with no
headroom limit, instead of a value that depends on how small the epsilon is
relative to `p`.

For test coverage, the cases that would have caught both of the above:

- `p = 4` (and ideally `p = 5`) on an all-zero slice — past the epsilon's range
- `p = 0.5` on a slice that mixes exact zeros with non-zeros and has a non-zero
  norm — the `self == 0` case
- comparing the non-degenerate rows against `torch.linalg.vector_norm(...).backward()`
  rather than only asserting `isfinite` and `== 0` on the zero row

One other thing worth flagging: the new test is under `tests/ttnn/nightly/`, and
the checks currently reported on this PR are lint and security scans only, so as
far as I can see it hasn't actually been run yet.

**Comment by mohmedmm** [comment]
Good catch on the underflow limit and p < 1 zero-elements. Replaced the epsilon with SFPU masking in moreh_norm_backward_kernel using mul_tiles_and_mask_tile_to_cb keyed on dfb_sign_obj, which zeroes out dx at x == 0 and for zero-norm slices without any numerical limits. Updated test cases to validate p in [0.5, 5.0] against PyTorch autograd.

**Comment by mohmedmm** [comment]
Could a maintainer with access trigger the Wormhole hardware CI (or provide cloud dev instance access) so we can validate this on silicon? Both underflow and p < 1 test vectors are ready.
